File: Python/2023maj/rgb.py
# 5. Feladat
def hatar(sor, elteres):
    for i in range(len(adat[sor]) - 1):
        kulonbseg = abs(adat[sor][i][2] - adat[sor][i+1][2])
        if kulonbseg > elteres:
            return True

    return False


if __name__ == "__main__":
    # 1. Feladat: beolvasás
    adat = []

    with open("kep.txt", "r", encoding="utf-8") as file:
        for sor in file:
            adat.append([])
            for idx, szam in enumerate(sor.strip().split(" ")):
                if idx % 3 == 0:
                    adat[-1].append([int(szam)])
                else:
                    adat[-1][-1].append(int(szam))

    # 2. Feladat
    print("2. feladat:")
    print("Kérem egy képpont adatait!")

    sor = int(input("Sor: ")) - 1
    oszlop = int(input("Oszlop: ")) - 1
    print(f"A képpont színe RGB({adat[sor][oszlop][0]},{adat[sor][oszlop][1]},{adat[sor][oszlop][2]})")

    # 3. Feladat
    print("3. feladat:")

    vilagos_pontok = 0
    for sor in adat:
        for pixel in sor:
            if sum(pixel) > 600:
                vilagos_pontok += 1

    print(f"A világos képpontok száma: {vilagos_pontok}")

    # 4. Feladat
    print("4. feladat:")

    legsotetebb_pont = 3 * 255
    for sor in adat:
        for pixel in sor:
            if sum(pixel) < legsotetebb_pont:
                legsotetebb_pont = sum(pixel)

    print(f"A legsötétebb pont RGB összege: {legsotetebb_pont}")

    print("A legsötétebb pixelek színe:")
    for sor in adat:
        for pixel in sor:
            if sum(pixel) == legsotetebb_pont:
                print(f"RGB({pixel[0]},{pixel[1]},{pixel[2]})")

    # 6. Feladat
    print("6. feladat:")

    i = 0
    while not hatar(i, 10):
        i += 1
    print(f"A felhő legfelső sora: {i + 1}")

    i = len(adat) - 1
    while not hatar(i, 10):
        i -= 1
    print(f"A felhő legalsó sora: {i + 1}")

File: go/2023maj/rgb.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

type Pixel struct {
    red int
    green int
    blue int
}

func (p Pixel) sum() int {
    return p.red + p.blue + p.green
}

func main()  {
    // 1. Feladat: beolvasás
    file, _ := os.Open("../../input/2023maj/kep.txt")
    defer file.Close()
    scanner := bufio.NewScanner(file)

    pixels := make([][]Pixel, 0)

    for scanner.Scan() {
        line := strings.Split(scanner.Text(), " ")
        pixels = append(pixels, make([]Pixel, 0))
        for i := 0; i < len(line); i+=3 {
            red, _ := strconv.Atoi(line[i])
            green, _ := strconv.Atoi(line[i+1])
            blue, _ := strconv.Atoi(line[i+2])
            pixels[len(pixels)-1] = append(pixels[len(pixels)-1], Pixel{red, green, blue})
        }
    }

    // 2. Feladat
    fmt.Println("Kérem egy képpont adatait!")
    var input string;

    fmt.Print("Sor: ")
    fmt.Scanln(&input)
    sor, _ := strconv.Atoi(strings.TrimSpace(input))

    fmt.Print("Oszlop: ")
    fmt.Scanln(&input)
    oszlop, _ := strconv.Atoi(strings.TrimSpace(input))

    fmt.Printf("A képpont színe RGB(%v, %v, %v)\n",
        pixels[sor-1][oszlop-1].red,
        pixels[sor-1][oszlop-1].green,
        pixels[sor-1][oszlop-1].blue,
    )

    // 3. Feladat
    vilagos := 0
    for _, s := range pixels {
        for _, p := range s {
            if p.sum() > 600 {
                vilagos++
            }
        }
    }

    fmt.Printf("3. feladat:\nVilágos képpontok száma: %v\n", vilagos)

    // 4. Feladat
    min := math.MaxInt
    for _, s := range pixels {
        for _, p := range s {
            if m := p.sum(); m < min {
                min = m
            }
        }
    }

    fmt.Printf("4. feladat:\nA legsötétebb pont RGB összege: %v\n", min)
    fmt.Println("A legsötétebb pixelek színe:")

    for _, s := range pixels {
        for _, p := range s {
            if p.sum() == min {
                fmt.Printf("RGB(%v, %v, %v)\n", p.red, p.green, p.blue)
            }
        }
    }

    // 6. Feladat
    var legfelso int;
    var legalso int;

    for i := 0; i < len(pixels); i++{
        if hatar(&pixels, i, 10) {
            legfelso = i + 1
            break
        }
    }

    for i := len(pixels)-1; i > 0; i--{
        if hatar(&pixels, i, 10) {
            legalso = i + 1
            break
        }
    }

    fmt.Printf("6. feladat:\nA felhő legfelső sora: %v\nA felhő legalsó sora: %v\n", legfelso, legalso)
}

// 5. Feladat 
func hatar(pixels *[][]Pixel, sor int, elteres int) bool {
    for i := 0; i < len((*pixels)[sor])-1; i++{
        if int(math.Abs(float64((*pixels)[sor][i].blue - (*pixels)[sor][i+1].blue))) > elteres {
            return true
        }
    }    
    return false
}
